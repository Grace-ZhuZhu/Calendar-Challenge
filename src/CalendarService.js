import _ from 'lodash';

const STARTING_HOUR = 9;
const WORKING_HOURS = 8;
const MINUTES_IN_ONE_HOUR = 60;

function getNewTimeSlot(event) {
    return {
        events: [event],
        start: event.start,
        end: event.end,
    }
}

export function getSortedEvents(events) {
    return events.sort((a, b) => a.start - b.start);
}

export function getMergedTimeBlocks(events) {
    const sortedEvents = getSortedEvents(events);

    return sortedEvents.reduce((result, event) => {
        if (!result.length) {
            result.push(getNewTimeSlot(event));
            return result;
        }

        const lastSlot = result[result.length - 1];

        if (event.start < lastSlot.end) {
            lastSlot.events.push(event);
            lastSlot.end = Math.max(lastSlot.end, event.end);
        } else {
            result.push(getNewTimeSlot(event))
        }

        return result;
    }, []);
}

export function getDisplayText(hour) {
    return hour < 10 ? `0${hour}` : `${hour}`;
}

export function getHour(time) {
    return Math.floor(time / MINUTES_IN_ONE_HOUR);
}

export function getTimeFromHour(hour) {
    return hour * MINUTES_IN_ONE_HOUR;
}

export function getMinutes(time) {
    return time % MINUTES_IN_ONE_HOUR;
}

export function getTimeDisplay(time) {
    const hour = getHour(time) + STARTING_HOUR;
    const minutes = getMinutes(time);

    return `${getDisplayText(hour)}:${getDisplayText(minutes)}`;
}

export function getMinutesToHourRatio(minutes) {
    return minutes / MINUTES_IN_ONE_HOUR;
}

export function getPercentage(ratio) {
    return `${ratio * 100}%`;
}

export function getEventOffset(eventStart, blockStart) {
    const ratio = getMinutesToHourRatio(eventStart - blockStart);
    return getPercentage(ratio);
}

export function getEventDisplay(event) {
    return `${getTimeDisplay(event.start)}-${getTimeDisplay(event.end)} EVENT`;
}

export function getEventHeightInPercentage(event) {
    const duration = event.end - event.start;
    const heightRatio = getMinutesToHourRatio(duration);
    return getPercentage(heightRatio);
}

export function getEventWidthInPercentage(numberOfEvents) {
    const widthRatio = 1/numberOfEvents;
    return getPercentage(widthRatio);
}

export function getEventForDisplay(event, blockStart, numberOfEvents) {
    return {
        displayText: getEventDisplay(event),
        height: getEventHeightInPercentage(event),
        width: getEventWidthInPercentage(numberOfEvents),
        topOffset:  getEventOffset(event.start, blockStart),
    };
}

export function getBlockOffset(time) {
    const minutes = getMinutes(time);
    const ratio = getMinutesToHourRatio(minutes);
    return getPercentage(ratio) ;
}

export function getEventBlocksForDisplay(timeBlocks) {
    return timeBlocks.map(block => {
        const {start, events} = block;
        const numberOfEvents =  events.length;

        return {
            startingHour: getHour(start),
            events: events.map(eve => getEventForDisplay(eve, start, numberOfEvents)),
            topOffset: getBlockOffset(start)
        }
    });
}

export function getHoursWithEventBlocks(events) {
    const timeBlocks = getMergedTimeBlocks(events);
    const eventsBlocks = getEventBlocksForDisplay(timeBlocks);

    return _.times(WORKING_HOURS).map(hour => {
        return {
            displayText: getTimeDisplay(getTimeFromHour(hour)),
            eventsBlock: eventsBlocks.find(block => block.startingHour === hour) || {},
        }
    });
}
