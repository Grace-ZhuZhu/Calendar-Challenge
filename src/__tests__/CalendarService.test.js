import * as Service from '../CalendarService';

const EVENT_1 = { start: 30, end: 90 };
const EVENT_2 = { start: 150, end: 210 };
const EVENT_3 = { start: 180, end: 210 };

const EVENTS = [
	EVENT_3,
	EVENT_2,
	EVENT_1,
];

const SORTED_EVENTS = [
	EVENT_1,
	EVENT_2,
	EVENT_3,
]

const EVENT_BLOCKS = [
	{
		start: 30,
		end: 90,
		events: [EVENT_1]
	},
	{
		start: 150,
		end: 210,
		events: [EVENT_2, EVENT_3]
	},
];

describe('getSortedEvents', () => {
	it('sort events by starting time', () => {
		const result = Service.getSortedEvents(EVENTS);

		expect(result).toEqual(SORTED_EVENTS);
	})
})

describe('getMergedTimeBlocks', () => {
	it('merges overlapping events in the same time block', () => {
		const result = Service.getMergedTimeBlocks(EVENTS);

		expect(result).toEqual(EVENT_BLOCKS);
	})
})

describe('getDisplayText', () => {
	it('returns the display text of 1 digit number of time', () => {
		const result = Service.getDisplayText(1);

		expect(result).toBe('01');
	})

	it('returns the display text of 2 digits number of time', () => {
		const result = Service.getDisplayText(11);

		expect(result).toBe('11');
	})
})

describe('getHour', () => {
	it('return the number of hours from the input time', () => {
		const result = Service.getHour(61);

		expect(result).toBe(1);
	})
})

describe('getTimeFromHour', () => {
	it('returns number of minutes from the input hour', () => {
		const result = Service.getTimeFromHour(1);

		expect(result).toBe(60);
	})
})

describe('getMinutes', () => {
	it('returns the minutes part of hh:mm format from input time', () => {
		const result = Service.getMinutes(60);

		expect(result).toBe(0);
	})
})

describe('getTimeDisplay', () => {
	it('returns display text of hh:mm format from input time', () => {
		const result = Service.getTimeDisplay(30);

		expect(result).toBe('09:30');
	})
})

describe('getMinutesToHourRatio', () => {
	it('returns the ration of the input time to the minutes of one hour', () => {
		const result = Service.getMinutesToHourRatio(30);

		expect(result).toBe(0.5);
	})
})

describe('getPercentage', () => {
	it('returns the percentage text of a ratio', () => {
		const result = Service.getPercentage(1.5);

		expect(result).toBe('150%');
	})
})

describe('getEventOffset', () => {
	it('returns the offset of the event start and block start in percentage of hour', () => {
		const result = Service.getEventOffset(60, 30);

		expect(result).toBe('50%');
	})
})

describe('getEventDisplay', () => {
	it('returns the display of an event', () => {
		const result = Service.getEventDisplay({ start: 30, end: 60 });

		expect(result).toBe('09:30-10:00 EVENT');
	})
})

describe('getEventHeightInPercentage', () => {
	it('returns the duration of an event in percentage of hour, as the height of an event item', () => {
		const result = Service.getEventHeightInPercentage({ start: 30, end: 60 });

		expect(result).toBe('50%');
	})
})

describe('getEventWidthInPercentage', () => {
	it('returns width of an event item', () => {
		const result = Service.getEventWidthInPercentage(4);

		expect(result).toBe('25%');
	})
})

describe('getEventForDisplay', () => {
	it('returns width of an event item', () => {
		const result = Service.getEventForDisplay({ start: 30, end: 60 }, 0, 2);

		expect(result).toEqual({
			displayText: "09:30-10:00 EVENT",
			height: "50%",
			topOffset: "50%",
			width: "50%"
		});
	})
})

describe('getBlockOffset', () => {
	it('returns the offset of the block start to the hour start in percentage of hour', () => {
		const result = Service.getBlockOffset(150);

		expect(result).toBe('50%');
	})
})

describe('getEventBlocksForDisplay', () => {
	it('returns the offset of the block start to the hour start in percentage of hour', () => {
		const result = Service.getEventBlocksForDisplay(EVENT_BLOCKS);

		console.log(JSON.stringify(result))

		expect(result).toMatchSnapshot();
	})
})

describe('getHoursWithEventBlocks', () => {
	it('returns the working hours with event blocks attached to the hour within which they start', () => {
		const result = Service.getHoursWithEventBlocks(EVENTS);

		expect(result).toMatchSnapshot();
	})
})