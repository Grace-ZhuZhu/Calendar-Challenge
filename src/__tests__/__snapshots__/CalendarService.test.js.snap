// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getEventBlocksForDisplay returns the offset of the block start to the hour start in percentage of hour 1`] = `
Array [
  Object {
    "events": Array [
      Object {
        "displayText": "09:30-10:30 EVENT",
        "height": "100%",
        "topOffset": "0%",
        "width": "100%",
      },
    ],
    "startingHour": 0,
    "topOffset": "50%",
  },
  Object {
    "events": Array [
      Object {
        "displayText": "11:30-12:30 EVENT",
        "height": "100%",
        "topOffset": "0%",
        "width": "50%",
      },
      Object {
        "displayText": "12:00-12:30 EVENT",
        "height": "50%",
        "topOffset": "50%",
        "width": "50%",
      },
    ],
    "startingHour": 2,
    "topOffset": "50%",
  },
]
`;

exports[`getHoursWithEventBlocks returns the working hours with event blocks attached to the hour within which they start 1`] = `
Array [
  Object {
    "displayText": "09:00",
    "eventsBlock": Object {
      "events": Array [
        Object {
          "displayText": "09:30-10:30 EVENT",
          "height": "100%",
          "topOffset": "0%",
          "width": "100%",
        },
      ],
      "startingHour": 0,
      "topOffset": "50%",
    },
  },
  Object {
    "displayText": "10:00",
    "eventsBlock": Object {},
  },
  Object {
    "displayText": "11:00",
    "eventsBlock": Object {
      "events": Array [
        Object {
          "displayText": "11:30-12:30 EVENT",
          "height": "100%",
          "topOffset": "0%",
          "width": "50%",
        },
        Object {
          "displayText": "12:00-12:30 EVENT",
          "height": "50%",
          "topOffset": "50%",
          "width": "50%",
        },
      ],
      "startingHour": 2,
      "topOffset": "50%",
    },
  },
  Object {
    "displayText": "12:00",
    "eventsBlock": Object {},
  },
  Object {
    "displayText": "13:00",
    "eventsBlock": Object {},
  },
  Object {
    "displayText": "14:00",
    "eventsBlock": Object {},
  },
  Object {
    "displayText": "15:00",
    "eventsBlock": Object {},
  },
  Object {
    "displayText": "16:00",
    "eventsBlock": Object {},
  },
]
`;
